### PR TITLE
chore: internal core refactorings

### PR DESCRIPTION
**Pull request by alexandrebodin**
### What does it do?

Remove loadedStrapi and init everything but admin in constructor

### CODE REVIEW

File: packages/core/content-releases/server/src/utils/index.ts [hunk 1]
```diff
@@ -1,16 +1,16 @@
-import type { UID, Data } from '@strapi/types';
+import type { UID, Data, Core } from '@strapi/types';
 
 export const getService = (
   name: 'release' | 'release-validation' | 'scheduling' | 'release-action' | 'event-manager',
-  { strapi } = { strapi: global.strapi }
+  { strapi }: { strapi: Core.Strapi }
```
Comment (Marc-Roig) [inline packages/core/content-releases/server/src/utils/index.ts:5]: this was assigned to the global strapi by default, couldn't this be an issue?

File: packages/core/core/src/Strapi.ts [hunk 2]
```diff
@@ -166,23 +142,20 @@ class Strapi extends Container implements Core.Strapi {
...
-      .add('plugins', registries.plugins(this))
-      .add('custom-fields', registries.customFields(this))
-      .add('apis', registries.apis(this))
-      .add('sanitizers', registries.sanitizers())
-      .add('validators', registries.validators())
       .add('query-params', createQueryParamService(this))
       .add('content-api', createContentAPI(this))
       .add('auth', createAuth())
-      .add('models', registries.models());
+      .add('server', () => createServer(this))
```
Comment (Marc-Roig) [inline packages/core/core/src/Strapi.ts:148]: would it make sense to have a type registry for the strapi containers? 
As far as I know doing strapi.get('containerName') does not return a typed value(?)

Comment (alexandrebodin) [reply packages/core/core/src/Strapi.ts:148]: Yes we will need to figure out a nice way to type it. Usually with di you type at loading but I don't like it, with the shortcut we maintain the all access with the types for now

File: packages/core/content-releases/server/src/utils/index.ts (hunk 1 above)
Comment (alexandrebodin) [reply packages/core/content-releases/server/src/utils/index.ts:5]: This was not a good pattern & an ensured this is a required param & is passed everywhere now

**Comment by alexandrebodin** [comment]
@Marc-Roig  & @Convly I'm not going all the way for now so we can get this merged and iterate on the providers & cleanup as a second phase

File: api-tests/utils/builder-helper.ts [hunk 3]
```diff
@@ -55,7 +55,7 @@ export const createTestSetup = async (
   bootstrapBuilder(resources, builder);
   await builder.build();
 
-  const strapi = (await createStrapiInstance()) as LoadedStrapi;
+  const strapi = (await createStrapiInstance()) as Core.Strapi;
```
Comment (Convly) [inline api-tests/utils/builder-helper.ts:58]:
```suggestion
  const strapi = await createStrapiInstance();
```

**Comment by innerdvations** [comment]
I'm trying to figure out exactly how to explain this and I'm not sure I'm getting it quite right, so bear with me 😆

The main purpose of LoadedStrapi for me was to ensure via a build error that the method couldn't be called unless bootstrap had run, meaning I had the logger and database available with content types loaded. Now it seems like all that logic will have to move in to the methods themselves with a check for strapi.isLoaded

**Comment by alexandrebodin** [comment]
> I'm trying to figure out exactly how to explain this and I'm not sure I'm getting it quite right, so bear with me 😆
> 
> The main purpose of LoadedStrapi for me was to ensure via a build error that the method couldn't be called unless bootstrap had run, meaning I had the logger and database available with content types loaded. Now it seems like all that logic will have to move in to the methods themselves with a check for strapi.isLoaded

Actually all accessors are using the container which will ensure the instance is created if it was registered. Meaning your code will error instead of receiving and undefined service so you don't have to check if you receive it it means it was instantiated successfuly

**Comment by alexandrebodin** [comment]
> > I'm trying to figure out exactly how to explain this and I'm not sure I'm getting it quite right, so bear with me 😆
> > The main purpose of LoadedStrapi for me was to ensure via a build error that the method couldn't be called unless bootstrap had run, meaning I had the logger and database available with content types loaded. Now it seems like all that logic will have to move in to the methods themselves with a check for strapi.isLoaded
> 
> Actually all accessors are using the container which will ensure the instance is created if it was registered. Meaning your code will error instead of receiving an undefined service so you don't have to check if you receive it it means it was instantiated successfuly. 

If services need special intilization code to be run that's a responsibility of the service itself not the service container

**Comment by innerdvations** [comment]
> > > I'm trying to figure out exactly how to explain this and I'm not sure I'm getting it quite right, so bear with me 😆
> > > The main purpose of LoadedStrapi for me was to ensure via a build error that the method couldn't be called unless bootstrap had run, meaning I had the logger and database available with content types loaded. Now it seems like all that logic will have to move in to the methods themselves with a check for strapi.isLoaded
> > 
> > 
> > Actually all accessors are using the container which will ensure the instance is created if it was registered. Meaning your code will error instead of receiving an undefined service so you don't have to check if you receive it it means it was instantiated successfuly.
> 
> If services need special intilization code to be run that's a responsibility of the service itself not the service container

Yeah, but I mean now we don't have a way via build error to say "all Strapi service initialization must have completed to use this", we only have strapi.isLoaded.

And if we drop down to service specific handling of initialization code, database doesn't currently even provide an `isInitialized`, so even though I'm able to do `strapi.get('db')` and get the database to use I have no way of knowing if it's actually ready to be used or not. I can just try it and see if it works. Being able to specify with types that we have everything ready and available is pretty useful especially when working with things inside the core services or bootstrap itself.

**Comment by alexandrebodin** [comment]
> > > > I'm trying to figure out exactly how to explain this and I'm not sure I'm getting it quite right, so bear with me 😆
> > > > The main purpose of LoadedStrapi for me was to ensure via a build error that the method couldn't be called unless bootstrap had run, meaning I had the logger and database available with content types loaded. Now it seems like all that logic will have to move in to the methods themselves with a check for strapi.isLoaded
> > > 
> > > 
> > > Actually all accessors are using the container which will ensure the instance is created if it was registered. Meaning your code will error instead of receiving an undefined service so you don't have to check if you receive it it means it was instantiated successfuly.
> > 
> > 
> > If services need special intilization code to be run that's a responsibility of the service itself not the service container
> 
> Yeah, but I mean now we don't have a way via build error to say "all Strapi service initialization must have completed to use this", we only have strapi.isLoaded.
> 
> And if we drop down to service specific handling of initialization code, database doesn't currently even provide an `isInitialized`, so even though I'm able to do `strapi.get('db')` and get the database to use I have no way of knowing if it's actually ready to be used or not. I can just try it and see if it works. Being able to specify with types that we have everything ready and available is pretty useful especially when working with things inside the core services or bootstrap itself.

I share the feeling of wanting the types to tell you if sth is ready to be used or not, but it's not possible. it just tells you if sth might be defined or not. Which is really different. You need runtime checks to ensure sth is ready to use regardless of typescript typings.

For ex 
```
db.query('xxx').find();
```

The issue is not to know if db is defined but if it has finished intializing. The type of the DB instance should still be the same you should never have 2 types for the same class. What you want can be achieve in two ways: 

1. pretty poor DX if you have to wrap every single service that has an init phase like that.
db.onReady(( dbQueryInstance) => {
 // make query with a type and remove db.query from the root
})

2. Have db.query() wait until the db is actually ready to execute the query or timeouts. 

The code could be staticaly defined and still break if you don't order the runtime execution correctly. 

You would never provider multiple types for your Database class with a Partial version for when it's ready to make queries or not, the same applies to Strapi or any framework Application.

We were merely faking all this but it was just shutting down typescript not giving us real guarantees.

**Comment by innerdvations** [comment]
> Have db.query() wait until the db is actually ready to execute the query or timeouts.

Yeah, that would be great, that's what I'll look into then if you don't get to it first 😆

File: packages/core/core/src/providers/registries.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+import { loadApplicationContext } from '../loaders';
+import * as draftAndPublishSync from '../migrations/draft-publish';
+
+export default defineProvider({
+  init(strapi) {
+    strapi
+      .add('content-types', () => registries.contentTypes())
+      .add('components', () => registries.components())
+      .add('services', () => registries.services(strapi))
+      .add('policies', () => registries.policies())
```
Comment (Marc-Roig) [inline packages/core/core/src/providers/registries.ts:14]: stupid question alert, 
does it make sense to keep the concept of a registry and a provider here? What makes content types require a registry but not crons?

Comment (alexandrebodin) [reply packages/core/core/src/providers/registries.ts:14]: Not sure I understand 🤔  
Crons have a registry but it's inside the cron service. and not in a shared registries folder.

Ideally I want to make those registries a bit smarter and move them into servies (like crons) 
Just don't want to stall too much that PR

**Review by Convly — APPROVED** [review]

**Pull request merged**